src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,599 @@
+'''
+backends/gdrive.py - this file is part of S3QL.
+
+Copyright © 2017 Isaac Aymerich <isaac.aymerich@gmail.com>
+
+This work can be distributed under the terms of the GNU GPLv3.
+'''
+
+
+from ..logging import logging # Ensure use of custom logger class
+from ..inherit_docstrings import (copy_ancestor_docstring,prepend_ancestor_docstring, ABCDocstMeta)
+from .common import (AbstractBackend, DanglingStorageURLError, NoSuchObject,retry,retry_generator,AuthenticationError)
+import apiclient.discovery
+import apiclient.http
+import time
+import base64
+import hashlib
+import httplib2
+import ssl
+import http
+import socket
+import io
+import json
+import oauth2client.client
+import tempfile
+
+log = logging.getLogger(__name__)
+logging.getLogger("googleapiclient.discovery").setLevel(logging.WARNING)
+
+
+
+# Maximum number of keys that can be deleted at once
+MAX_KEYS = 1000
+credentials= None
+mainFolder= None
+class Backend(AbstractBackend, metaclass=ABCDocstMeta):
+    '''
+    A backend that stores data in Google Drive.
+    '''
+
+    known_options = set()
+    MIME_TYPE_BINARY = 'application/octet-stream'
+    MIME_TYPE_FOLDER = 'application/vnd.google-apps.folder'
+    TOKEN_URI = 'https://accounts.google.com/o/oauth2/token'
+    USER_AGENT = 'S3QL-Backend'
+
+    MAX_PROPERTY_SIZE = 124
+    MAX_REF_CHUNK_SIZE = 100
+    VALUE_BASE64 = 'b64:'
+    VALUE_REF = 'ref:'
+    
+
+    def __init__(self, storage_url, login, password, options):
+        '''Initialize local backend
+        Login and password are ignored.
+        '''
+        # Unused argument
+        #pylint: disable=W0613
+
+        super().__init__()
+        self.login = login
+        self.password = password
+        client_secret_and_refresh_token = self.password.split(':')
+        if len(client_secret_and_refresh_token) != 2:
+            raise AuthenticationError("Invalid Password format, must be  secretApplication:refreshToken")
+        self.client_secret = client_secret_and_refresh_token[0]
+        self.refresh_token = client_secret_and_refresh_token[1]
+        self.options = options
+        self.delete_map_state = dict()
+
+        # get google drive service
+        global credentials        
```

Comment:
No global variables please. What if you want to have multiple backend instances that use different credentials?

Reply:
I wanted to avoid every thread refresh google refresh token by itself so this way the token only needed to refresh on one of them and all remaining can get the refreshed token. any idea how to do it better?

Reply:
Yes. Take a look at how the Google Storage backend does it (look for `_refresh_lock`). I think one of the swift backends also has similar functionality.